### PR TITLE
server: fix regular user can create isolated network without sourcenat

### LINKED ISSUE #5627: Regular user can create isolated networks without Source Nat
According to code, Regular user can create Guest Isolated Source Nat enabled network only

https://github.com/apache/cloudstack/blob/4.16.0.0-RC20211025T0851/server/src/main/java/com/cloud/network/NetworkServiceImpl.java#L1363-L1369

```
        // Regular user can create Guest Isolated Source Nat enabled network only
        if (_accountMgr.isNormalUser(caller.getId()) && (ntwkOff.getTrafficType() != TrafficType.Guest
                || ntwkOff.getGuestType() != Network.GuestType.Isolated && areServicesSupportedByNetworkOffering(ntwkOff.getId(), Service.SourceNat))) {
            throw new InvalidParameterValueException(
                    String.format("Regular users can only create a network from network offerings having traffic type [%s] and network type [%s] with a service [%s] enabled.", TrafficType.Guest,
                            Network.GuestType.Isolated, Service.SourceNat.getName()));
        }
```

However regular user can still create Isolated networks without source nat.

##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
Networking
~~~

##### CLOUDSTACK VERSION

~~~
4.16.0.0-RC2
~~~

##### STEPS TO REPRODUCE

~~~

~~~

##### EXPECTED RESULTS

~~~

~~~

##### ACTUAL RESULTS

~~~

~~~

### PR DESCRIPTION
**Pull request by weizhouapache**
### Description

This PR fixes #5627

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [x] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by weizhouapache** [comment]
@blueorangutan package

**Review by DaanHoogland — APPROVED** [review]
clgtm (but would benefit in terms of readability if it was put in a separate method)

File: server/src/main/java/com/cloud/network/NetworkServiceImpl.java [hunk 1]
```diff
@@ -1360,12 +1360,22 @@ public Network createGuestNetwork(CreateNetworkCmd cmd) throws InsufficientCapac
...
+                        TrafficType.Guest, GuestType.Isolated, Service.SourceNat.getName()));
+            } else if (ntwkOff.getGuestType() == GuestType.L2) {
+                s_logger.debug(String.format("Creating a network from network offerings having traffic type [%s] and network type [%s].",
+                        TrafficType.Guest, GuestType.L2));
+            } else {
+                throw new InvalidParameterValueException(
+                        String.format("Regular users can only create an %s network with a service [%s] enabled, or a %s network.",
+                                GuestType.Isolated, Service.SourceNat.getName(), GuestType.L2));
+            }
         }
```
[the remark this answers is not in the document]
Comment (weizhouapache) [reply server/src/main/java/com/cloud/network/NetworkServiceImpl.java]: @DaanHoogland done. thanks for review

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

File: server/src/main/java/com/cloud/network/NetworkServiceImpl.java [hunk 2]
```diff
@@ -1457,6 +1454,23 @@ public Network createGuestNetwork(CreateNetworkCmd cmd) throws InsufficientCapac
...
+        }
+        if (ntwkOff.getGuestType() == GuestType.Isolated && areServicesSupportedByNetworkOffering(ntwkOff.getId(), Service.SourceNat)) {
+            s_logger.debug(String.format("Creating a network from network offerings having traffic type [%s] and network type [%s] with a service [%s] enabled.",
+                    TrafficType.Guest, GuestType.Isolated, Service.SourceNat.getName()));
+        } else if (ntwkOff.getGuestType() == GuestType.L2) {
+            s_logger.debug(String.format("Creating a network from network offerings having traffic type [%s] and network type [%s].",
+                    TrafficType.Guest, GuestType.L2));
+        } else {
+            throw new InvalidParameterValueException(
+                    String.format("Regular users can only create an %s network with a service [%s] enabled, or a %s network.",
```
Comment (GabrielBrascher) [inline server/src/main/java/com/cloud/network/NetworkServiceImpl.java:1469]: Great log messages :+1:

**Review by GabrielBrascher — APPROVED** [review]
Code LGTM

**Review by sureshanaparti — APPROVED** [review]
LGTM

**Pull request merged**